src/MaxText/input_pipeline/_grain_data_processing.py

```diff
@@ -66,7 +68,35 @@ def create_dataset_from_pattern(pattern):
       source = grain.ArrayRecordDataSource(files)
       return grain.MapDataset.source(source)
 
-    if ";" in data_file_pattern:
+    # Handle mixture config with named datasets, allows flexibility in recovering checkpoints
+    if mixture_config_path:
+      with open(mixture_config_path, "r", encoding="utf-8") as f:
+        mixture_config = json.load(f)
+
+      paths = [config["path"] for config in mixture_config.values()]
+      weights = [float(config["weight"]) for config in mixture_config.values()]
+
+      executor = futures.ThreadPoolExecutor(max_workers=grain_data_source_max_workers)
+      dataset_list = list(executor.map(create_dataset_from_pattern, paths))
+      executor.shutdown(wait=True)
+
+      datasets_dict = dict(zip(mixture_config.keys(), dataset_list))
+
+      for name, ds in datasets_dict.items():
+        if shuffle:
+          ds = ds.shuffle(seed=shuffle_seed)
```

Comment:
It seems we are shuffle --> repeat --> shard, have you thought about shard --> shuffle --> repeat for efficiency?